Allows solution of integral(sqrt(a*x + b) / x, x) (#10847)

lindsayad opened the pull request:
Additions to manualintegrate.py:
- SOverXBLtZeroRule: handles condition when b < 0
- SOverXAXGtZeroRule: handles condition when b > 0 and a \* x > 0
- SOverXAXLtZeroRule: handles condition when b > 0 and a \* x < 0
- In pull_out_polys function with _parts_rule, added additional or condition
  that allows integration by parts to be performed in conjunction with
  non-integer powers

Fixes #10847

Commit:
Allows solution of integral(sqrt(a*x + b) / x, x)

Additions to manualintegrate.py:
 - SOverXBLtZeroRule: handles condition when b < 0
 - SOverXAXGtZeroRule: handles condition when b > 0 and a * x > 0
 - SOverXAXLtZeroRule: handles condition when b > 0 and a * x < 0
 - In pull_out_polys function with _parts_rule, added additional or condition
   that allows integration by parts to be performed in conjunction with
   non-integer powers

Fixes #10847

lindsayad commented:
Please let me know if there's additional work (or complete re-work!) that needs to be done in order for this to go in.

lindsayad commented:
Can someone tell this contributing newbie why the different tests are failing?

jksuom commented:
It looks like the code in manualintegrate is running in a loop. You could investigate by testing one new rule at a time.

lindsayad commented:
Yay, all tests passing now :-)

asmeurer commented:
Please add tests for the new functionality.

jksuom commented:
I think that a useful manualintegrate rule should be able to deal with more than just one specific type of integrand. With this rule we get the result
```
>>> manualintegrate(sqrt(2*x + 3)/x, x)
Piecewise((2*sqrt(2*x + 3) - 2*sqrt(3)*acoth(sqrt(3)*sqrt(2*x + 3)/3), 2*x > 0), (2*sqrt(2*x + 3) - 2*sqrt(3)*atanh(sqrt(3)*sqrt(2*x + 3)/3), 2*x < 0))
```
but variants are not evaluated:
```
>>> manualintegrate(sqrt(2*x + 3)/(x + 1), x)
Integral(sqrt(2*x + 3)/(x + 1), x)
>>> manualintegrate(sqrt(2*x + 3)/2*x, x)
Integral(x*sqrt(2*x + 3), x)/2
```
A useful rule would cover all integrands of the form `f(x, sqrt(a*x + b))` where `f(x, y)` is a rational fraction two variables. It would work by setting `u = sqrt(a*x + b)` and then expressing `x` in terms of `u` turning the integrand into a rational function of `u` that can be integrated by standard means.

Commit:
Added rational denominator rule.

Commit:
Prevents substep=None.

Commit:
Managed to remove None exceptions for issue #9462.

Commit:
Fixed tests so tests pass with new rational denom rule.

Commit:
Added tests for issue #9462.

Commit:
Added #10847 tests demonstrating new rational and root power integration capability.

Commit:
Implemented rational root fraction capability and added tests.

Commit:
Cleaned up code and added better names.

Commit:
Totally removed old ArctanRule. Renamed QDRules to ArctanRule, ArccothRule, and ArctanhRule.

lindsayad commented:
Alright, I added code that I believe implements the root fraction capability @jksuom is looking for. I also added checks that prevent the errors reported in #9462. Additionally, the ArctanRule has been re-implemented to be both more general (has some matching) but yet to cover only the conditions for which the ArctanRule is valid. Consequently, tests using the old ArctanRule have been re-written.

asmeurer commented:
Looks like there are test failures.

jksuom commented on sympy/integrals/manualintegrate.py:
```diff
@@ -342,7 +342,10 @@ def _parts_rule(integrand, symbol):
         integrand = integrand.together()
-        polys = [arg for arg in integrand.args if arg.is_polynomial(symbol)]
+        polys = [arg for arg in integrand.args if (arg.is_polynomial(symbol) or \
+                                                   (isinstance(arg, sympy.Pow) and \
```
What is the purpose of these additions? According to the name this method aims to extract polynomials. But these additions would include powers with noninteger and negative exponents.

lindsayad replied:
It allows integration of expressions like x**(3/2) * log(x) and x**(-3) * log(x), which were previously not integrable with manualintegrate

jksuom replied:
Why are such expressions treated as polynomials? There should probably be a separate section of code for them.

lindsayad replied:
I feel like the noninteger and negative exponents should be lumped in with the polynomials. I interpret the 'A' in 'LIATE' as meaning algebraic expressions which is a larger set of expressions than polynomials. Having that check does not break any of the existing manualintegrate tests and it adds functionality. As opposed to breaking into multiple sections of code (which based on some experimentation could break some tests), I propose simply renaming pull_out_polynomials to pull_out_algebraic or something like that. In my opinion, that would be more consistent with the LIATE concept.

jksuom replied:
It may be useful to include expressions of algebraic type, but I'd expect them to be of the form `polynomial**rational`. If the base contains transcendental functions or the exponent is irrational, the power cannot be considered algebraic. Powers with exponent -1 are also algebraic (even rational). It is not clear why they should not be allowed.

lindsayad replied:
I'm going to be thinking about this, but any suggestions for writing a test similar to is_polynomial, but instead tests is_algebraic?

lindsayad replied:
Like is_polynomial, it would be nice if `(x**a).is_rational()` evaluates to `True`, but `(a**x).is_rational()` evaluates to `False` where `x` is a symbol and `a` is a constant. I think that would be most useful for the integral rules. You are right, that `a = -1` should be allowed

jksuom replied:
Possibly the methods `is_rational_function` or `is_algebraic_expr` could be used.

lindsayad replied:
I went ahead and replaced `is_polynomial` with `is_algebraic_expr`. Now I'm getting all the functionality that I wanted.

Commit:
Prevent root_mul_rule from trying to handle cases with multiple root
functions that have different arguments. Without this check, get a recursion
error when evaluating test_issue_6253 in test_integrals.

Commit:
Added tests displaying new functionality for integrating by parts.

Now integration by parts can handle algebraic expressions with negative and
non-integer exponents in addition to regular polynomials.

lindsayad commented:
I keep having these comparison issues where expressions that appear to be exactly equivalent compare falsely, example illustrated below:

```
In[80]: %paste
    result = manualintegrate(x**(3/2) * log(x), x)
    desired_result = 0.4*x**2.5*log(x) - 0.16*x**2.5
    print(result)
    print(desired_result)
    print(result == desired_result)
    assert result == desired_result
## -- End pasted text --
0.4*x**2.5*log(x) - 0.16*x**2.5
0.4*x**2.5*log(x) - 0.16*x**2.5
False
Traceback (most recent call last):
  File "/home/lindsayad/miniconda2/envs/python3SympyDev/lib/python3.5/site-packages/IPython/core/interactiveshell.py", line 2885, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-80-5bc67a132f24>", line 6, in <module>
    assert result == desired_result
AssertionError
```

Is there a way to overcome this problem?

Commit:
Converted float to Rational for successful rational fraction power test.

lindsayad commented:
@asmeurer's suggestion of converting to Rational fixed the above problem. Thank you!

Commit:
Pursing is_algebraic_expr

Commit:
Finish transforming pull_out_polys to pull_out_algebraic

Additionally, added a check that prevents nan calculations stemming from
divisions by zero in evaluating CyclicPartsRule

lindsayad commented:
That's a fun little failed test. Looks like I'm getting penalized because manualintegrate can now do a little more work. In the failing test in test_prde (test issue #10798), this integrand gets passed to manualintegrate: `(log(1 - 1/y)/2 - log(1 + 1/y)/2)/y`

Without my changes this evaluates to `Integral(log(1 - 1/y)/y, y)/2 - Integral(log(1 + 1/y)/y, y)/2`. Or in other words, it does nothing.

With my changes, it is evaluted to: `log(y)*log(1 - 1/y)/2 - log(y)*log(1 + 1/y)/2 - Integral(log(y)/(y**2*(1 - 1/y)), y)/2 - Integral(log(y)/(y**2*(1 + 1/y)), y)/2`

Thoughts on how to proceed?

Commit:
Added simplification in _parts_rule that increases robustness of some
comparisons.

Added test for manualintegrate(log(y)/(y**2*(1 - 1/y)), y) showing increased
integration capability. Unsolved integral is recognizable as a polylog using
meijer integration

lindsayad commented:
Alright I added some simplification in _parts_rule that allows further integration of `(log(1 - 1/y)/2 - log(1 + 1/y)/2)/y`. Now:

`manualintegrate((log(1 - 1/y)/2 - log(1 + 1/y)/2)/y)` 

yields

 `log(y)*log((y - 1)/y)/2 - log(y)*log((y + 1)/y)/2 - log(y)*log(y - 1)/2 + log(y)*log(y + 1)/2 + Integral(log(y - 1)/y, y)/2 - Integral(log(y + 1)/y, y)/2`

The latter two terms are now recognizable by meijer integration such that:

`integrate(1/(1-(x*y)**2), (x, 0, 1), y)` 

yields:

`-Piecewise((I*pi*log(y) - polylog(2, y), Abs(y) < 1), (-I*pi*log(1/y) - polylog(2, y), Abs(1/y) < 1), (-I*pi*meijerg(((), (1, 1)), ((0, 0), ()), y) + I*pi*meijerg(((1, 1), ()), ((), (0, 0)), y) - polylog(2, y), True))/2 - log(y)*log(1 - 1/y)/2 + log(y)*log(1 + 1/y)/2 + log(y)*log(y - 1)/2 - log(y)*log(y + 1)/2 + I*pi*log(y)/2 - polylog(2, y*exp_polar(I*pi))/2`

This differs in form from the current test result, but they should be mathematically equivalent.

lindsayad commented:
I suppose that it could be made cleaner if manualintegrate recognized the initial integrand that it's passed as having a solution containing a polylog

lindsayad commented:
The hanging test in `test_integrals --slow` is test_issue_7130. When I test this on my own machine, 50% of the time it passes in around 15 seconds, while 50% of the time it hangs. I don't really know what's going on there

Commit:
Edited test_prde test_issue_10798 result.

lindsayad commented:
May I have some review on this? @jksuom @asmeurer? Even if not all the tweaks are immediately desirable, I think that there are some easy, important ones that could go in right away. For example, the check before CyclicPartsRule that prevents a potential division by zero and a dead-end integration path.

jksuom commented:
It seems to me that no real progress is made in the code producing this expression

`log(y)*log((y - 1)/y)/2 - log(y)*log((y + 1)/y)/2 - log(y)*log(y - 1)/2 + log(y)*log(y + 1)/2 + Integral(log(y - 1)/y, y)/2 - Integral(log(y + 1)/y, y)/2`

since the integrated part simplifies to zero. In fact, I was expecting that you would modify that code.

lindsayad commented:
Thanks for your reply. If I make the assumption: `y = symbols('y', positive=True)`, then the simplification to zero occurs as you mention, e.g. `print(sympy.simplify(manualintegrate((log(1 - 1/y)/2 - log(1 + 1/y)/2)/y, y)))` yields `Integral(log(y - 1)/y, y)/2 - Integral(log(y + 1)/y, y)/2`. And also, the result of test_issue_10798 in test_prde is:

 `integrate(1/(1-(x*y)**2), (x, 0, 1), y)` = `I*pi*log(y)/2 + polylog(2, exp_polar(I*pi)/y)/2 - polylog(2, exp_polar(2*I*pi)/y)/2 + pi**2/4` 

which differs from the result for the master branch only by the constant `pi**2/4`. So in the end you are definitely correct in saying that my PR doesn't make any additional progress on the integrand `(log(1 - 1/y)/2 - log(1 + 1/y)/2)/y`. However, I don't think I can really make any more progress in `manualintegrate` with that particular integrand unless I introduced a rule that recognized a polylog candidate, which is beyond what I was hoping to achieve in this particular PR. However, I could look at adding it in if you would really like. (But maybe in a different PR?) My initial goal really was to be able to address issue #10847, which I then tried to expand to integrands of the form `f(x, sqrt(a*x + b))` as you suggested. I think I've achieved that, but please let me know your thoughts.

jksuom commented:
Your discussion of the log-integrand led me to believe, apparently mistakenly, that it was going to be a central point. I also think that polylog need not be a part of this PR. Otherwise this looks promising, but there may be a problem. `test_integrals` has produced no output.

jksuom commented:
It seems that the slow test is hanging in `test_issue_7130`.

lindsayad commented:
Yes, you are right. As represented by the fact that this slow test passed for python3.5 and failed for python2.7, in my own testing this `test_issue_7130` hangs about 50% of the time and runs successfully 50% of the time. I am ignorant as to why this seemingly probabilistic behavior is occurring. If multiple threads were running, I would think I might be having a race condition, but these calculations are purely sequential. Right? What other reason might there be for this probabilistic behavior? This is a new realm of learning for me :-) Could it be a caching issue similar to [this?](https://github.com/sympy/sympy/issues/7702)

Commit:
Zoning in on some good simplifications.

Commit:
Realized problem is repeated integration of algebraic with non-positive-integer power times an exponential function. Without using special functions, I believe this integration will go on forever, leading to the hanging test.

Commit:
Add check that prevents infinite looping of algebraic_expr and exp. Tests passing locally.

lindsayad commented:
@jksuom All tests are passing now

jksuom commented on sympy/integrals/manualintegrate.py:
```diff
@@ -436,14 +440,17 @@ def make_second_step(steps, integrand):
         else:
-            return integral_steps(integrand, symbol)
+            if integral_steps(integrand, symbol):
+                return integral_steps(integrand, symbol)
```
There should be only one call of `integral_steps`. Its return value can be saved and only then tested.

jksuom commented:
This looks good. Only, calling twice a major function such as `integral_steps` should be avoided.

Commit:
Eliminate expensive double calling of integral_steps.

lindsayad commented:
Thank you @jksuom. I believe I have gotten rid of the all the multiple calls of major functions like `integral_steps` in the newest commit.

jksuom commented:
Thank you for your work. I'll merge in 24 h if there a no more comments.

lindsayad commented:
Awesome. Thanks for all the helpful review. I am very pleased to have made my first contribution to SymPy!
